python/paddle/fluid/layers/nn.py

```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
             "spatial_scale": spatial_scale
         })
     return pool_out
+
+
+def dice_loss(input, label, num_classes, reduce_dim=[1, 2, 3], epsilon=0.00001):
+    """
+    **Dice loss Layer**
+    Dice loss for comparing the similarity of two batch of data,
+    usually be used for binary image segmentation i.e. labels are binary.
+    The value between 0 to 1, 1 means totally match.
+
+    Args:
+        input (Variable): The predictions with shape [batch_size, ... , num_classes].
+        label (Variable): The groud truth with shape [batch_szie, ... , 1].
+        num_classes (integer): The number of target classes.
+        reduce_dim (list<int>): The dimensions to be reduced. It should be set to input.shape[1:].
+                                Default: [1,2,3]
+        epsilon (float): The epsilon will be added to the numerator and denominator.
+                         If both input and label are empty, it makes sure dice is 1.
+                         Default: 0.00001
+
+    Returns:
+        dice_loss (Variable): The dice loss.
+
+    Examples:
+            predictions = fluid.layers.softmax(x)
+            loss = fluid.layers.dice_loss(input=predictions, label=label, 2)
+    """
+    helper = LayerHelper('dice_loss', **locals())
+    out = helper.create_tmp_variable(dtype=input.dtype)
+    label = one_hot(label, depth=num_classes)
+    inse = reduce_sum(input * label, dim=reduce_dim)
+
+    dice_denominator = reduce_sum(
+        input, dim=reduce_dim) + reduce_sum(
+            label, dim=reduce_dim)
+    dice_score = (inse * 2 + epsilon) / (dice_denominator + epsilon)
+    return reduce_mean(dice_score)
```

Comment:
It seems the returned loss should be 1 - dice_score from https://github.com/pytorch/pytorch/issues/1249 .

Reply:
It may not be necessary to add epsilon to the numerator

Reply:
Fixed. Thx.

Reply:
> It may not be necessary to add epsilon to the numerator

Should be necessary since prediction [0,0,0,0] and target [0,0,0,0] expect a perfect score.